thesis-de/content-de.tex

```diff
@@ -2715,8 +2731,123 @@ \section{Realisierung Grafana}
+mergen.
+\tm%
+
+\section{Continuous Integration}
+Continuous Integration beschreibt eine Vorgehensweise in der
+Softwareentwicklung. Hierbei wird für jede Änderung eine zuvor definierte
+Testmatrix durchlaufen, um sicherzustellen, dass die Software weiterhin intakt
+ist. Dies setzt vorraus, das Tests for dem eigentlichen Code entwickelt werden
+(Test Driven Development) oder parallel zur Entwicklung des Codes. Ersteres
+stellt sicher, das ausschlieslich das implementiert wird, das gefordert ist.
+In den Tests wird definiert wie eine Software zu funktionieren hat, danach
+wird so lange am Code entwickelt bis die Tests erfolgreich sind. Wenn man die
+Tests parallel zum Code entwickelt, ist es möglich, dass der Code mehr als die
+geforderte Funktionalität aufweist.
+
+Im Rahmen des Projekts werden Tests immer parallel zum Code entwickelt. Test
+Driven Development erfordert ein enormes Fachwissen über die genutzten
+Frameworks für Tests. Dieses lässt sich, aufgrund der komplexität, nicht
+während des Projektes aneignen.
+\tm%
+
+\subsection{Travis}
+Travis ist eine Platform für Continuous Integration. Sie ist kostenlos nutzbar
+und stellt verschiedene virtuelle Maschinen und Container bereit, in denen
+Tests ausgeführt werden können. GitHub kann bei neuen Pull Requests eine
+Benachrichtigung an Travis schicken. Daraufhin startet Travis eine neue VM und
+klont das entsprechende GitHub Repository und den \texttt{Feature Branch}.
+Danach wird eine definierte Testmatrix ausgeführt. Travis schickt eine
+Benachrichtigung an GitHub, nachdem die Tests beendet wurden. Somit ist in der
+Weboberfläche ersichtlich, ob die Tests erfolgreich waren oder nicht.
+
+Dieses Verfahren ermöglicht es, eine reproduzierbare Testumgebung zu nutzen.
+Diese wird immer auotmatisch gestartet. Es ist nicht mehr notwendig, jeden Test
+lokal auszuführen. Travis stellt VMs bereit, in denen nur das nötigste
+installiert ist. Dadurch ist sichergestellt, dass keine \glslink{Side
+Effect}{Side Effects} auftreten. Bei lokalen Tests tritt dies häufig auf. Der
+Entwickler hat diverse Bibliotheken installiert, welche von der entwickelten
+Software benötigt werden.  Innerhalb der entwickelten Software wird eine
+Abhängigkeit nicht definiert, es funktioniert aber trotzdem, da die Bibliothek
+zufällig vorhanden ist.  Innerhalb von Travis fällt dies aber auf. Außerdem ist
+es jedem Teammitglied möglich, die Travis Logs von überall aus zu begutachten.
+Alternativ müsste jeder Entwickler die Tests selbst lokal ausführen.
+\tm%
+
+\section{Linter und Syntax Validator}
+
+\section{Unit Tests}
+Für die Tests und Qualitätsicherung wurden unter anderem Unit Tests verwendet.
+Unit Tests sind Software Tests und werden immer dann verwendet, wenn jeweils
+die Kernfunktion von einem Quellcode geprüft werden soll. Das Framework hierzu
+ist von rspec entwickelt worden. Rspec besitzt bereits schon eine Reihe an
+Test, welche für die im Projekt verwendeten Profile entweder erweitert oder neu
```

Comment:
rspec besitzt keine tests. rspec stellt eine DSL (steht im glossar) bereit in der man tests schreiben kann. die puppet module die wir nutzen, z.b. grafana, besitzen schon tests. diese wurden erweitert.